pennylane/_qubit_device.py

```diff
@@ -280,20 +282,29 @@ def execute(self, circuit, **kwargs):
 
                 if qml.math._multi_dispatch(r) == "jax":  # pylint: disable=protected-access
                     r = r[0]
-                elif not isinstance(r[0], dict):
+                elif no_counts:
                     # Measurement types except for Counts
                     r = qml.math.squeeze(r)
-                if isinstance(r, (np.ndarray, list)) and r.shape and isinstance(r[0], dict):
-                    # This happens when measurement type is Counts
-                    results.append(r)
+
+                if not no_counts:
+
+                    # This happens when at least one measurement type is Counts
+                    for result_group in r:
+                        if isinstance(result_group, list):
```

Comment:
This confused me for a while, but now I think I get what's happening.  While this works, we probably need to try and extract and simplify a lot of this code around here, eventually.

Reply:
Yes, a refactor as a whole could be done for example once we have a more refined return type implementation.